Doc updates (#542)
funcx-faas/funcX: docs/endpoints.rst

Endpoints
=========

> An endpoint is a persistent service launched by the user on a compute to serve as a conduit for

Reviewer: ```suggestion
An endpoint is a persistent service launched by the user on a compute system to serve as a conduit for
```

> executing functions on that computer. funcX supports a range of target systems enabling

Reviewer: ```suggestion
executing functions on that computer. funcX supports a range of target systems, enabling
```

an endpoint to be deployed on a laptop, the login node of a campus cluster, a cloud instance,
or a Kubernetes cluster, for example.

The endpoint requires outbound network connectivity. That is, it must be able to connect to
funcX at `funcx.org <https://api2.funcx.org/v2>`_.

To install the funcX endpoint agent software ::

     $ python3 -m pip install funcx_endpoint

.. note::

   Please note that the funcx endpoint is only supported on Linux and MacOS.

> After installing funcX endpoint you be able to deploy new endpoints and interact

Reviewer: ```suggestion
After installing funcX endpoint, you be able to deploy new endpoints and interact
```

with existing endpoints using the `funcx-endpoint` command.

First time setup
----------------

You will be required to authenticate the first time you run `funcx-endpoint`.
If you have authenticated previously, the endpoint will cache access tokens in
the local configuration file.

funcX requires authentication in order to associate

> endpoints with users and ensure only authorized users can run tasks on that endpoint. As part of this step

Reviewer: ```suggestion
endpoints with users and ensure only authorized users can run tasks on that endpoint. As part of this step,
```

we request access to your identity and Globus Groups.

> To get started you will first want to configure a new endpoint.  ::

Reviewer: ```suggestion
To get started, you will first want to configure a new endpoint.  ::
```

  $ funcx-endpoint configure

Once you've run this command, a directory will be created at `$HOME/.funcx` and a set of default configuration files will be generated.

You can also set up auto-completion for the `funcx-endpoint` commands in your shell, by using the command ::

  $ funcx-endpoint --install-completion [zsh bash fish ...]

Configuring an Endpoint
----------------------------

> After creating an endpoint, a configuration file will be created at `$HOME/.funcx/config.py`. This file contains

Reviewer 2: @kylechard I think we should remove this bit completely. We no longer rely on this file.

information regarding the endpoint, such as the owner username and email address.
It also includes information about the endpoint's broker, a concept used when the endpoint controls execution
on other nodes (e.g., in a cluster deployment).

> .. note:: If your endpoint will manage execution on other nodes you must set the endpoint address such that workers can communicate with the endpoint.

Reviewer: ```suggestion
.. note:: If your endpoint will manage execution on other nodes, you must set the endpoint address such that workers can communicate with the endpoint.
```

Reviewer: should this be "so that"?

    'endpoint_address': address_by_hostname(),
  }

FuncX endpoints act as gateways to diverse computational resources, including clusters, clouds,
supercomputers, and even your laptop. To make the best use of your resources, the endpoint must be
configured to match the capabilities of the resource on which it is deployed.

FuncX provides a Python class-based configuration model that allows you to specify the shape of the
resources (number of nodes, number of cores per worker, walltime, etc.) as well as allowing you to place
limits on how funcX may scale the resources in response to changing workload demands.

To generate the appropriate directories and default configuration template, run the following command::

  $ funcx-endpoint configure <ENDPOINT_NAME>

> This command will create a profile for your endpoint in `$HOME/.funcx/<ENDPOINT_NAME>/` and will instantiate a

Reviewer: In a different part of this PR, Yadu suggested that this file is no longer used.

  $ funcx-endpoint start <ENDPOINT_NAME>

.. note:: If the ENDPOINT_NAME is not specified, a default endpoint named "default" is started.

Starting an endpoint will perform a registration process with funcX.
The registration process provides funcX with information regarding the endpoint.
The endpoint establishes three outbound
ZeroMQ channels to funcX (on the three ports returned during registration)
to retrieve tasks, send results, and communicate command information.

Once started, the endpoint uses a daemon process to run in the background.

.. note:: If the endpoint was not stopped correctly previously (e.g., after a computer restart when the endpoint was running), the endpoint directory will be cleaned up to allow a fresh start

.. warning:: Only the owner of an endpoint is authorized to start an endpoint. Thus if you register an endpoint

> using a different identity and try to start an endpoint owned by another identity, it will fail.

Reviewer: ```suggestion
using one identity and try to start an endpoint owned by another identity, it will fail.
```

Endpoints can be the following states:

> * **Initialized**: the endpoint has been created, but not started

Reviewer: ```suggestion
* **Initialized**: The endpoint has been created, but not started
```

>   following configuration and not registered with the `funcx service`

Reviewer: ```suggestion
  following configuration and is not registered with the `funcx service`.
```

> * **Running**: the endpoint is active and available for executing  functions

Reviewer: ```suggestion
* **Running**: The endpoint is active and available for executing  functions.
```

> * **Stopped**: the endpoint was stopped by the user. It is not running

Reviewer: ```suggestion
* **Stopped**: The endpoint was stopped by the user. It is not running
```

  and therefore, cannot service any functions. It can be started again without issues.

> * **Disconnected**: the endpoint disconnected unexpectedly. It is not running

Reviewer: ```suggestion
* **Disconnected**: The endpoint disconnected unexpectedly. It is not running
```

  and therefore, cannot service any functions. Starting this endpoint will first invoke
  necessary endpoint cleanup, since it was not stopped correctly previously.

Container behaviors and routing
-------------------------------

> The funcX endpoint can run functions using indepenent Python processes or optionally

Reviewer: ```suggestion
The funcX endpoint can run functions using independent Python processes or optionally
```

inside containers. funcX supports various container technologies (e.g., docker and singularity)
and different routing mechanisms for different use cases.

Raw worker processes (`worker_mode=no_container`):

* Hard routing: All worker processes are of the same type "RAW". It this case, the funcx endpoint simply routes tasks to any available worker processes. This is the default mode of a funcx endpoint.

> * Soft routing: It is the same as hard routing.

Reviewer: ```suggestion
* Soft routing: This is the same as hard routing.
```

Reviewer: This doesn't make sense to me

> * Soft routing: It is the same as hard routing.

Reviewer: ```suggestion
* Soft routing: For raw worker processes, this is the same as hard routing.
```

Reviewer: would it make more sense to say soft routing is not supported in this case?  I don't think I understand how soft and hard can be the same

Reviewer: or maybe hard and soft aren't really very good terms for the distinction we are trying to make here

Reviewer 2: We struggled a bit with naming here. I think we should come back to this in a separate pass to avoid a hold up.
